MAID-2574: fix direct connections

povilasb opened the pull request:
* Added direct connect integration test that used to fail
* Implemented race condition during direct connections fix
* Fixed exchange_data example that used to prematurely drop Service object, thus connections were dropped too

povilasb committed:
refactor/tests: have separate tests cases for each protocol

Better isolates tests - better reporting on what's working and not.

povilasb committed:
fix/examples: don't prematurely drop Service

Updates example to keep Service and Listener objects alive while example is running. Otherwise connections will be dropped and example won't work as expected.

povilasb committed:
test/service: add direct connection test when only one peer is accessible directly

This test is failing for now and shows a race condition in direct connections.
Peer1 is accessible directly, Peer2 is not. Peer2 initiates connection and after 2 seconds so Peer1.
Connection fails, because by the time Peer2 sends connect message to Peer1, Peer1 is not expecting it.

povilasb committed:
fix/connect: fix race condition in direct connections

When new unknown connections comes in, cache it until it's used. When we attempt to connect to a peer, we initially check the connection cache, if it's not already there.

povilasb committed:
refactor/demux: extract functions

maidsafe-highfive commented:
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @canndrew (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

povilasb committed:
feat/demux: use timeouts for incoming connections

If some random connection comes in, we store it. If it's used, ok, if not - don't store it cached for too long.
